File: airbnb-net/ClientApp/src/services/AutocompleteService.ts
import {HttpCustomClient} from "./HttpCustomClient";

export class AutocompleteService {
    http: HttpCustomClient;
    baseUrl: string;
    private constructor() {
        this.http = HttpCustomClient.getInstance();
        this.baseUrl = "/autocomplete";
    }
    
    private static instance: AutocompleteService;
    public static getInstance(): AutocompleteService {
        if (!AutocompleteService.instance) {
            AutocompleteService.instance = new AutocompleteService();
        }
        return AutocompleteService.instance;
    }
    
    async getAutocompleteData(query: string, limit: number) {
        return await this.http.Get<string[]>(`${this.baseUrl}/?query=${query}&limit=${limit}`);
    }
    
    async getGeocode(query: string) {
        return await this.http.Get<{Latitude: number, Longitude: number}>(`${this.baseUrl}/geocode/?query=${query}`);
    }
}

File: ApplicationLogic/Services/AutocompleteService.cs
using ApplicationCommon.GeospatialUtilities;
using ApplicationLogic.Options;
using BingMapsRESTToolkit;
using Microsoft.Extensions.Options;

namespace ApplicationLogic.Services;

public class AutocompleteService : IAutocompleteService
{
    private readonly BingMapsConnectionOptions _bingMapsConnectionOptions;

    public AutocompleteService(IOptions<BingMapsConnectionOptions> bingMapsConnectionOptions)
    {
        _bingMapsConnectionOptions = bingMapsConnectionOptions.Value;
    }

    public async Task<IEnumerable<string>> GetAutocomplete(string query, int limit)
    {
        var request = new AutosuggestRequest()
        {
            Query = query,
            MaxResults = limit,
            IncludeEntityTypes = new List<AutosuggestEntityType>() { AutosuggestEntityType.Address, AutosuggestEntityType.Place},
            UserLocation = new Coordinate(47.60357, -122.32945),
            BingMapsKey = _bingMapsConnectionOptions.BingMapsKey
        };
        
        var response = await request.Execute();

        if (response == null ||
            response.ResourceSets == null ||
            response.ResourceSets.Length <= 0 ||
            response.ResourceSets[0].Resources == null ||
            response.ResourceSets[0].Resources.Length <= 0) return new List<string>();
        var results = (Autosuggest)response.ResourceSets[0].Resources[0];
        return results != null ? results.Value.Select(r => r.Address.FormattedAddress): throw new Exception("Unable to get autocomplete results.");
    }

    public async Task<Coordinates> GetGeocode(string query)
    {
        var request = new GeocodeRequest()
        {
            Query = query,
            BingMapsKey = _bingMapsConnectionOptions.BingMapsKey
        };

        var response = await request.Execute();
        
        if (response == null ||
            response.ResourceSets == null ||
            response.ResourceSets.Length <= 0 ||
            response.ResourceSets[0].Resources == null ||
            response.ResourceSets[0].Resources.Length <= 0) throw new Exception("Unable to get geocode results.");
        var boundingBox = response.ResourceSets[0].Resources[0].BoundingBox;
        var coordinates = new Coordinates((boundingBox[0] + boundingBox[2]) / 2, (boundingBox[1] + boundingBox[3]) / 2);
        return coordinates;
    }
}